File: ruby/sieve.rb
def sieve(n)
  nums = (0..n).to_a
  nums[0] = nil
  nums[1] = nil
  4.step(n, 2) do |i|
    nums[i] = nil
  end
  3.step( (n ** 0.5).floor, 2) do |i|
    next unless nums[i]
    
    (i * i).step(n, i + i) do |j|
      nums[j] = nil
    end
  end
  nums.compact!
end

def is_prime(n)
  return false if n <= 1
  return true if n <= 3
  sieve( (n ** 0.5).floor ).each do |i|
    return false if (n % i == 0)
  end
  return true
end

File: perl/sieve.pl
#!/usr/bin/perl

use strict;
use warnings;

sub sieve {
    my $n = shift;
    my @nums;

    for my $i (0 .. $n) {
        $nums[$i] = $i;
    }

    for (my $i = 3; $i <= int(sqrt($n)); $i = $i + 2) {
        next unless $nums[$i];
        
        for (my $j = $i * $i; $j <= $n; $j = $j + $i * 2) {
            $nums[$j] = 0;
        }
    }

    return grep { $_ != 0 && $_ % 2 != 0; } @nums;
}

my $n = <STDIN>;
chomp($n);

my @primes = sieve(int($n));
print $#primes + 1, " ", $primes[$#primes], "\n";
